File: library/src/main/java/me/proxer/library/api/ProxerResponse.java
package me.proxer.library.api;

import com.squareup.moshi.Json;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.annotation.Nullable;

/**
 * @author Ruben Gees
 */
@SuppressWarnings("TypeParameterUnusedInFormals")
@ToString
@EqualsAndHashCode(onParam = @__({@Nullable}))
@NoArgsConstructor(access = AccessLevel.PRIVATE)
final class ProxerResponse<T> {

    @SuppressWarnings("unused")
    @Getter(AccessLevel.NONE)
    @Json(name = "error")
    private boolean error;

    @SuppressWarnings("unused")
    @Getter(AccessLevel.PACKAGE)
    @Json(name = "message")
    private String message;

    @SuppressWarnings("unused")
    @Getter(AccessLevel.PACKAGE)
    @Json(name = "code")
    private int code;

    @Nullable
    @SuppressWarnings("unused")
    @Getter(AccessLevel.PACKAGE)
    @Json(name = "data")
    private T data;

    boolean isSuccessful() {
        return !error;
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/ProxerResponse.kt
package me.proxer.library.internal

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.proxer.library.internal.adapter.NumberBasedBoolean

/**
 * @author Ruben Gees
 */
@JsonClass(generateAdapter = true)
internal data class ProxerResponse<T>(
    @field:NumberBasedBoolean @Json(name = "error") internal val error: Boolean,
    @Json(name = "message") val message: String,
    @Json(name = "code") val code: Int?,
    @Json(name = "data") val data: T?
) {

    val isSuccessful = !error
}
